qa/L0_lifecycle/lifecycle_test.py

```diff
@@ -2236,6 +2237,149 @@ def test_config_override(self):
                 except Exception as ex:
                     self.assertTrue(False, "unexpected error {}".format(ex))
 
+    def test_shutdown_dynamic(self):
+        model_shape = (1, 1)
+        input_data = np.ones(shape=(1, 1), dtype=np.float32)
+
+        inputs = [grpcclient.InferInput('INPUT0', model_shape, "FP32")]
+        inputs[0].set_data_from_numpy(input_data)
+
+        triton_client = grpcclient.InferenceServerClient("localhost:8001",
+                                                         verbose=True)
+        model_name = "custom_zero_1_float32"
+
+        # Send two requests as only requests held in scheduler are counted
+        # as in-flight (the first request is in execution)
+        def callback(user_data, result, error):
+            if error:
+                user_data.append(error)
+            else:
+                user_data.append(result)
+
+        # Currently the dynamic batcher will form payloads and place to
+        # instance queue in advance. The batcher doesn't track requests
+        # in the next stage so need to send more requests to saturate the
+        # queue.
+        request_count = 6
+        async_results = []
+        for _ in range(request_count):
+            triton_client.async_infer(model_name, inputs,
+                                      partial(callback, async_results))
+        time.sleep(1)
+
+        # Send signal to shutdown the server
+        os.kill(int(os.environ['SERVER_PID']), signal.SIGINT)
+
+        # Send more requests and should be rejected
+        try:
+            triton_client.infer(model_name, inputs)
+            self.assertTrue(False,
+                            "expected error for new inference during shutdown")
+        except InferenceServerException as ex:
+            self.assertIn("Scheduler has stopped accepting new inference",
+                          ex.message())
+
+        # Wait until the results are available in user_data
+        time_out = 30
+        while ((len(async_results) < request_count) and time_out > 0):
+            time_out = time_out - 1
+            time.sleep(1)
+
+        # Previous requests should succeed
+        for result in async_results:
+            if type(result) == InferenceServerException:
+                raise result
+            output_data = result.as_numpy('OUTPUT0')
+            self.assertTrue(np.allclose(output_data, input_data),
```

Comment:
An optional tidbit: I know we use assertTrue and np.allclose together in most places, but another option is Numpy's built-in function for this. It addresses edge cases (that may not be relevant for Triton) and provides clearer output. Sharing below, feel free to ignore/resolve this comment.

https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html